Subject: [PATCH] Fix "Failed to read row" while deleting .nomedia files

Use constant/offset query in onMediaScannerConnected()
to prevent android.process.media crash because of
possibly IllegalStateException in "Failed to read row"
from CursorWindow which caused by clients call
MediaProvider's deleting to directory with .nomedia.

diff --git a/src/com/android/providers/media/MediaProvider.java b/src/com/android/providers/media/MediaProvider.java
index 8dc853e..f5da832 100644
--- a/src/com/android/providers/media/MediaProvider.java
+++ b/src/com/android/providers/media/MediaProvider.java
@@ -3927,20 +3927,36 @@ public ScannerClient(Context context, SQLiteDatabase db, String path) {
 
         @Override
         public void onMediaScannerConnected() {
-            Cursor c = mDb.query("files", openFileColumns,
+            
+            Cursor c = null;
+            Cursor maxC = mDb.query("files", openFileColumns,
                     "_data >= ? AND _data < ?",
                     new String[] { mPath + "/", mPath + "0"},
                     null, null, null);
+
+            int pages = maxC.getCount()/1000;
             try  {
-                while (c.moveToNext()) {
-                    String d = c.getString(0);
-                    File f = new File(d);
-                    if (f.isFile()) {
-                        mScannerConnection.scanFile(d, null);
+                for (int i = 0; i < pages + 1; i++) {
+
+                    c = mDb.query("files", openFileColumns,
+                            "_data >= ? AND _data < ?",
+                            new String[] { mPath + "/", mPath + "0"},
+                            null, null, null, "" + i * 1000 + ",1000");
+                    
+                    Log.d(TAG, "c.getCount() = " + c.getCount());
+                    Log.d(TAG, "mPath = " + mPath);
+
+                    while (c.moveToNext()) {
+                        String d = c.getString(0);
+                        File f = new File(d);
+                        if (f.isFile()) {
+                            mScannerConnection.scanFile(d, null);
+                        }
                     }
                 }
-                mScannerConnection.disconnect();
             } finally {
+                mScannerConnection.disconnect();
+                IoUtils.closeQuietly(maxC);
                 IoUtils.closeQuietly(c);
             }
         }
